### PR TITLE
Improve Device Trust Ceremony mocking for tests

### PR DESCRIPTION
**Pull request by Joerger**
Replace global testing variables for device trust with pluggable ceremony interface, as recommended here - https://github.com/gravitational/teleport/pull/26862/files#r1205562481

### CODE REVIEW

**Comment by Joerger** [comment]
@codingllama Does this need to be backported to v11?

**Comment by codingllama** [comment]
> @codingllama Does this need to be backported to v11?

Nope, there's no device trust prior to v12.

File: lib/devicetrust/authn/authn.go [hunk 1]
```diff
@@ -24,21 +24,25 @@ import (
 	"github.com/gravitational/teleport/lib/devicetrust/native"
 )
 
-// vars below are used to swap native methods for fakes in tests.
-var (
-	getDeviceCredential = native.GetDeviceCredential
-	collectDeviceData   = native.CollectDeviceData
-	signChallenge       = native.SignChallenge
-)
+type CeremonyI interface {
```
Comment (codingllama) [inline lib/devicetrust/authn/authn.go:27]: Let's drop the interface, or at least move the interface to lib/client. (See comments on lib/client.)

We don't want the interface here, it goes against the advice on https://github.com/golang/go/wiki/CodeReviewComments#interfaces.

File: lib/devicetrust/authn/authn.go [hunk 2]
```diff
@@ -24,21 +24,25 @@ import (
...
+
+func NewCeremony() *Ceremony {
+	return &Ceremony{
+		GetDeviceCredential: native.GetDeviceCredential,
+		CollectDeviceData:   native.CollectDeviceData,
+		SignChallenge:       native.SignChallenge,
+	}
+}
+
+func (c *Ceremony) Run(ctx context.Context, devicesClient devicepb.DeviceTrustServiceClient, certs *devicepb.UserCertificates) (*devicepb.UserCertificates, error) {
```
Comment (codingllama) [inline lib/devicetrust/authn/authn.go:45]:
```suggestion
// RunCeremony performs the client-side device authentication ceremony.
//
// Device authentication requires a previously registered and enrolled device
// (see the lib/devicetrust/enroll package).
//
// The outcome of the authentication ceremony is a pair of user certificates
// augmented with device extensions.
func (c *Ceremony) Run(ctx context.Context, devicesClient devicepb.DeviceTrustServiceClient, certs *devicepb.UserCertificates) (*devicepb.UserCertificates, error) {
```

File: lib/devicetrust/authn/authn.go [hunk 3]
```diff
@@ -24,21 +24,25 @@ import (
...
 
-// RunCeremony performs the client-side device authentication ceremony.
-//
-// Device authentication requires a previously registered and enrolled device
-// (see the lib/devicetrust/enroll package).
-//
-// The outcome of the authentication ceremony is a pair of user certificates
-// augmented with device extensions.
-func RunCeremony(ctx context.Context, devicesClient devicepb.DeviceTrustServiceClient, certs *devicepb.UserCertificates) (*devicepb.UserCertificates, error) {
+type Ceremony struct {
```
Comment (codingllama) [inline lib/devicetrust/authn/authn.go:31]:
```suggestion
// Ceremony is the device authentication ceremony.
// It takes the client role of
// [devicepb.DeviceTrustServiceClient.AuthenticateDevice].
type Ceremony struct {
```

File: lib/devicetrust/authn/authn.go [hunk 4]
```diff
@@ -24,21 +24,25 @@ import (
...
+	GetDeviceCredential func() (*devicepb.DeviceCredential, error)
+	CollectDeviceData   func() (*devicepb.DeviceCollectedData, error)
+	SignChallenge       func(chal []byte) (sig []byte, err error)
+}
+
+func NewCeremony() *Ceremony {
```
Comment (codingllama) [inline lib/devicetrust/authn/authn.go:37]:
```suggestion
// NewCeremony creates a new ceremony that delegates per-device behavior
// to lib/devicetrust/native.
// If you want to customize a [Ceremony], for example for testing purposes, you
// may create a configure an instance directly, without calling this method.
func NewCeremony() *Ceremony {
```

File: lib/devicetrust/authn/authn.go [hunk 5]
```diff
@@ -111,3 +115,14 @@ func RunCeremony(ctx context.Context, devicesClient devicepb.DeviceTrustServiceC
...
 	return newCerts, nil
 }
+
+// RunCeremony performs the client-side device authentication ceremony.
+//
+// Device authentication requires a previously registered and enrolled device
+// (see the lib/devicetrust/enroll package).
+//
+// The outcome of the authentication ceremony is a pair of user certificates
+// augmented with device extensions.
```
Comment (codingllama) [inline lib/devicetrust/authn/authn.go:125]:
```suggestion
// It is equivalent to calling `NewCeremony().Run()`.
```

File: lib/devicetrust/authn/authn_test.go [hunk 6]
```diff
@@ -57,30 +56,21 @@ func TestRunCeremony(t *testing.T) {
...
 
-			_, err := authn.RunCeremony(ctx, devices, test.certs)
+			_, err := ceremony.Run(ctx, devices, test.certs)
 			// A nil error is good enough for this test.
 			assert.NoError(t, err, "RunCeremony failed")
 		})
 	}
 }
 
-func resetNative() func() {
```
Comment (codingllama) [inline lib/devicetrust/authn/authn_test.go:73]: Nice, this is looking tidier already.

File: lib/client/api_login_test.go [hunk 7]
```diff
@@ -459,11 +459,8 @@ func TestTeleportClient_DeviceLogin(t *testing.T) {
 	t.Run("device login", func(t *testing.T) {
 		// We need this because the running standalone process is not Enterprise.
 		teleportClient.SetDTAttemptLoginIgnorePing(true)
-
-		// validatingRunCeremony checks the parameters passed to dtAuthnRunCeremony
-		// and returns validCerts on success.
 		var runCeremonyCalls int
-		validatingRunCeremony := func(_ context.Context, devicesClient devicepb.DeviceTrustServiceClient, certs *devicepb.UserCertificates) (*devicepb.UserCertificates, error) {
+		validatingRunCeremony := func(ctx context.Context, devicesClient devicepb.DeviceTrustServiceClient, certs *devicepb.UserCertificates) (*devicepb.UserCertificates, error) {
```
Comment (codingllama) [inline lib/client/api_login_test.go:463]: nit: param unused?

```suggestion
		validatingRunCeremony := func(_ context.Context, devicesClient devicepb.DeviceTrustServiceClient, certs *devicepb.UserCertificates) (*devicepb.UserCertificates, error) {
```

File: lib/client/api_login_test.go [hunk 8]
```diff
@@ -459,11 +459,8 @@ func TestTeleportClient_DeviceLogin(t *testing.T) {
 	t.Run("device login", func(t *testing.T) {
 		// We need this because the running standalone process is not Enterprise.
 		teleportClient.SetDTAttemptLoginIgnorePing(true)
-
-		// validatingRunCeremony checks the parameters passed to dtAuthnRunCeremony
```
Comment (codingllama) [inline lib/client/api_login_test.go:463]: Fix/keep the comment?

File: lib/client/api.go [hunk 9]
```diff
@@ -968,10 +968,10 @@ type TeleportClient struct {
 	// Useful to force flows that only typically happen on Teleport Enterprise.
 	dtAttemptLoginIgnorePing, dtAutoEnrollIgnorePing bool
 
-	// dtAuthnRunCeremony allows tests to override the default device
+	// dtAuthnCeremony allows tests to override the default device
 	// authentication function.
-	// Defaults to [dtauthn.RunCeremony].
-	dtAuthnRunCeremony dtAuthnRunCeremonyFunc
```
Comment (codingllama) [inline lib/client/api.go:974]: You don't really need to change any of this, we can just assign `NewCeremony().Run` to it.

File: lib/client/api.go [hunk 10]
```diff
@@ -3398,14 +3398,14 @@ func (tc *TeleportClient) DeviceLogin(ctx context.Context, certs *devicepb.UserC
 	}
 
 	// Allow tests to override the default authn function.
-	runCeremony := tc.dtAuthnRunCeremony
-	if runCeremony == nil {
-		runCeremony = dtauthn.RunCeremony
+	ceremony := tc.dtAuthnCeremony
+	if ceremony == nil {
+		ceremony = dtauthn.NewCeremony()
```
Comment (codingllama) [inline lib/client/api.go:3403]: As suggested above, keep all of this unchanged and do:

```go
		runCeremony = dtauthn.NewCeremony().Run
```

We don't really gain anything with the new interface and `mockCeremony`.

File: lib/devicetrust/authn/authn.go [hunk 11]
```diff
@@ -111,3 +115,14 @@ func RunCeremony(ctx context.Context, devicesClient devicepb.DeviceTrustServiceC
...
+func RunCeremony(ctx context.Context, devicesClient devicepb.DeviceTrustServiceClient, certs *devicepb.UserCertificates) (*devicepb.UserCertificates, error) {
```
Comment (codingllama) [inline lib/devicetrust/authn/authn.go:126]: Despite the comment above, it looks like this isn't used anymore, so we can also just drop it.

**Review by codingllama — COMMENTED** [review]
Thanks a lot for this, Brian.

Copying @strideynet, as this has direct impact on his ongoing PRs. I think we should do a matching refactor on lib/devicetrust/enroll, but I don't want to abuse of your goodwill, so Noah or myself can take it - let's just wrap this one so they both look the same.

File: lib/devicetrust/authn/authn.go [hunk 12]
```diff
@@ -26,23 +26,37 @@ import (
...
+
+// NewCeremony creates a new ceremony that delegates per-device behavior
+// to lib/devicetrust/native.
+// If you want to customize a [Ceremony], for example for testing purposes, you
+// may create a configure an instance directly, without calling this method.
+func NewCeremony() *Ceremony {
+	return &Ceremony{
+		GetDeviceCredential: native.GetDeviceCredential,
+		CollectDeviceData:   native.CollectDeviceData,
+		SignChallenge:       native.SignChallenge,
```
Comment (codingllama) [inline lib/devicetrust/authn/authn.go:48]: It doesn't look like we are initializing all fields here.

```suggestion
		GetDeviceCredential:          native.GetDeviceCredential,
		CollectDeviceData:            native.CollectDeviceData,
		SignChallenge:                native.SignChallenge,
		SolveTPMAuthnDeviceChallenge: native.SolveTPMAuthnDeviceChallenge,
		GetDeviceOSType:              native.GetDeviceOSType,
```

**Review by codingllama — COMMENTED** [review]
Looks good, barring that one outstanding comment.

**Review by strideynet — APPROVED** [review]

**Review by codingllama — APPROVED** [review]

**Pull request merged**